File: app/src/main/java/com/artemchep/horario/_new/fragments/SubjectTasksFragment.java
package com.artemchep.horario._new.fragments;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.artemchep.horario.R;
import com.artemchep.horario._new.activities.GalleryActivity;
import com.artemchep.horario._new.activities.SubjectActivity;
import com.artemchep.horario._new.adapters.AdapterModel;
import com.artemchep.horario._new.adapters.StreamAdapter;
import com.artemchep.horario._new.adapters.TasksAdapter;
import com.artemchep.horario.aggregator.Filter;
import com.artemchep.horario.database.models.SubjectInfo;
import com.artemchep.horario.database.models.SubjectTask;
import com.artemchep.horario.ui.activities.ChildActivity;
import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.FirebaseDatabase;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

/**
 * @author Artem Chepurnoy
 */
public class SubjectTasksFragment extends FragmentModel<SubjectTask> implements SubjectActivity.Page {

    private ChildEventListener mChildEventListener = new ChildEventListener() {
        @Override
        public void onChildAdded(DataSnapshot dataSnapshot, String s) {
            SubjectTask task = dataSnapshot.getValue(SubjectTask.class);
            task.key = dataSnapshot.getKey();
            getAggregator().put(task);
        }

        @Override
        public void onChildChanged(DataSnapshot dataSnapshot, String s) {
            SubjectTask task = dataSnapshot.getValue(SubjectTask.class);
            task.key = dataSnapshot.getKey();
            getAggregator().put(task);
        }

        @Override
        public void onChildRemoved(DataSnapshot dataSnapshot) {
            getAggregator().remove(dataSnapshot.getKey());
        }

        @Override
        public void onChildMoved(DataSnapshot dataSnapshot, String s) {
        }

        @Override
        public void onCancelled(DatabaseError databaseError) {
        }
    };
    private SubjectInfo mSubject;

    @Nullable
    @Override
    public View onCreateView(
            LayoutInflater inflater, @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.___fragment_subject_lessons, container, false);
    }

    @Override
    public void onStart() {
        super.onStart();
        getAggregator().reset();
        FirebaseDatabase.getInstance().getReference()
                .child("subjects").child("testing_is_fun").child("tasks")
                .addChildEventListener(mChildEventListener);
    }

    @Override
    public void onStop() {
        super.onStop();
        FirebaseDatabase.getInstance().getReference()
                .child("subjects").child("testing_is_fun").child("tasks")
                .removeEventListener(mChildEventListener);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void setSubject(@Nullable SubjectInfo subject) {
        mSubject = subject;
    }

    @Override
    public Filter<SubjectTask> onCreateFilter() {
        return new Filter<>();
    }

    @Override
    public Comparator<SubjectTask> onCreateComparator() {
        return new Comparator<SubjectTask>() {
            @Override
            public int compare(SubjectTask a, SubjectTask b) {
                return a.key.compareToIgnoreCase(b.key);
            }
        };
    }

    @Override
    public AdapterModel<SubjectTask, ? extends AdapterModel.ViewHolder<SubjectTask>> onCreateAdapter(List<SubjectTask> list) {
        return new TasksAdapter(this, list);
    }

    @Override
    public void onItemClick(View v, SubjectTask data) {
        super.onItemClick(v, data);

        Bundle args = new Bundle();
        args.putString(SubjectTaskFragment.EXTRA_SUBJECT_ID, "-KlnzthUb_8St7Ax0UT0");
        args.putString(SubjectTaskFragment.EXTRA_TASK_ID, "keykeykey");
        args.putString(SubjectTaskFragment.EXTRA_GROUP_ID, "keykeykey");
        args.putInt(SubjectTaskFragment.EXTRA_COLOR, mSubject.color);

        Intent intent = new Intent(getContext(), ChildActivity.class);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        intent.putExtra(ChildActivity.EXTRA_FRAGMENT_NAME, SubjectTaskFragment.class.getName());
        intent.putExtra(ChildActivity.EXTRA_FRAGMENT_ARGS, args);
        startActivity(intent);
    }

    @Override
    public void onFabClick(@NonNull View view) {
    }

    @Override
    public boolean hasFab() {
        return true;
    }

}

File: app/src/main/java/com/artemchep/horario/ui/fragments/SubjectTasksFragment.kt
package com.artemchep.horario.ui.fragments

import android.content.Context
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.StaggeredGridLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.artemchep.horario.Palette
import com.artemchep.horario.R
import com.artemchep.horario._new.activities.SubjectActivity
import com.artemchep.horario.database.data.Task
import com.artemchep.horario.database.models.ITask
import com.artemchep.horario.interfaces.FragmentHost
import com.artemchep.horario.ui.adapters.AdapterBase
import com.artemchep.horario.ui.utils.*
import com.artemchep.horario.ui.widgets.UserView
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import java.util.*

/**
 * @author Artem Chepurnoy
 */
class SubjectTasksFragment : FragmentStore<ITask>(), AdapterBase.OnItemClickListener, SubjectActivity.Page {

    private var color: Int = 0
    private lateinit var userId: String
    private lateinit var subjectId: String

    /** Reference to a list of the tasks of this subject */
    private lateinit var tasksQuery: Query

    override val filter: (ITask) -> Boolean = { true }
    override val comparator: Comparator<ITask> = compareBy(
            { it.timestamp },
            { it.id }
    )

    override fun onAttach(context: Context?) {
        arguments!!.run {
            userId = getString(EXTRA_USER_ID)
            subjectId = getString(EXTRA_SUBJECT_ID)
            color = getInt(EXTRA_COLOR, Palette.UNKNOWN)
        }

        val firestore = FirebaseFirestore.getInstance()
        tasksQuery = firestore.collection("subjects/$subjectId/tasks")
        super.onAttach(context)
    }

    override fun onCreateAdapter(list: MutableList<ITask>): Adapter = Adapter(context!!, list)

    override fun onCreateContentView(
            inflater: LayoutInflater, container: ViewGroup?,
            savedInstanceState: Bundle?): View {
        return inflater.inflate(R.layout.___fragment_subject_tasks, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        view.findViewById<RecyclerView>(R.id.recycler).run {
            // Show two columns of posts in
            // `large` interfaces
            layoutManager = if (isLarge) {
                StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.VERTICAL)
            } else LinearLayoutManager(activity)
        }
    }

    override fun onItemClick(view: View, pos: Int) {
        val data = adapter.getItem(pos)
        val args = Bundle().apply {
            putInt(EXTRA_COLOR, color)
            putString(EXTRA_USER_ID, userId)
            putString(EXTRA_SUBJECT_ID, subjectId)
            putString(EXTRA_TASK_ID, data.getKey())
            putParcelable(EXTRA_TASK, data)
        }

//        val host = activity as FragmentHost
//        host.fragmentShow(SubjectTaskFragment::class.java, args, FragmentHost.FLAG_AS_SECONDARY)
    }

    override fun onFabClick(view: View) {
        val args = Bundle().apply {
            putInt(EXTRA_COLOR, color)
            putString(EXTRA_USER_ID, userId)
            putString(EXTRA_SUBJECT_ID, subjectId)
        }

        val host = activity as FragmentHost
        host.fragmentShow(SubjectTaskEditFragment::class.java, args, FragmentHost.FLAG_AS_SECONDARY)
    }

    override fun hasFab(): Int = R.drawable.ic_plus_white_24dp

    override fun getQuery(): Query = tasksQuery

    override fun getModel(snapshot: DocumentSnapshot): ITask = Task.from(snapshot)!!

    /**
     * @author Artem Chepurnoy
     */
    class Adapter(private val context: Context, list: List<ITask>) : AdapterBase<ITask, Adapter.ViewHolder>(list) {

        /**
         * @author Artem Chepurnoy
         */
        class ViewHolder(adapter: Adapter, v: View) : AdapterBase.ViewHolder<Adapter>(adapter, v) {
            internal val userView: UserView = itemView.findViewById(R.id.user_view)
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            val inflater = LayoutInflater.from(parent.context)
            val v = inflater.inflate(R.layout.___item_task, parent, false)
            return ViewHolder(this, v)
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            val post = getItem(position)
            val res = context.resources

            holder.userView.modelId = post.author
        }

    }

}
